zerver/lib/events.py

```diff
@@ -1000,19 +1000,19 @@ def _draft_update_action(i: int) -> None:
     elif event["type"] == "stream":
         if event["op"] == "create":
             for stream in event["streams"]:
-                if not stream["invite_only"]:
-                    stream_data = copy.deepcopy(stream)
-                    if include_subscribers:
-                        stream_data["subscribers"] = []
-
-                    # We know the stream has no traffic, and this
-                    # field is not present in the event.
-                    #
-                    # TODO: Probably this should just be added to the event.
-                    stream_data["stream_weekly_traffic"] = None
-
-                    # Add stream to never_subscribed (if not invite_only)
-                    state["never_subscribed"].append(stream_data)
+                stream_data = copy.deepcopy(stream)
+                if include_subscribers:
+                    stream_data["subscribers"] = []
+
+                # We know the stream has no traffic, and this
+                # field is not present in the event.
+                #
+                # TODO: Probably this should just be added to the event.
+                stream_data["stream_weekly_traffic"] = None
```

Comment:
Since the stream creation event is sent when user gains access to a stream, so the `stream_weekly_traffic` can be non-null as well. I think we can include the `stream_weekly_traffic` data in the event only as mentioned in TODO. But I wanted to confirm do we want to include it for all the stream objects sent to the server like the objects sent in response of various endpoints and in `streams` field of register response. Would be nice to have it consistently sent in all the stream objects. 
Currently it is included only in subscription objects (i.e. stream details combined with subscription information like stream-level notification settings, etc.). Since the traffic is stream related information and does not depend on the user receiving the information, there should be no problem in sending it in the stream objects.

Reply:
I think the concern problem about sending it more broadly (or including it in events) is that it is potentially expensive to compute. I don't remember the details and could be wrong; I'm pretty sure it uses the StreamCount table so is likely not expensive to do for a single stream.

Reply:
So, the main function used is this -
```python
def get_streams_traffic(stream_ids: Set[int]) -> Dict[int, int]:
    stat = COUNT_STATS["messages_in_stream:is_bot:day"]
    traffic_from = timezone_now() - datetime.timedelta(days=28)

    query = StreamCount.objects.filter(property=stat.property, end_time__gt=traffic_from)
    query = query.filter(stream_id__in=stream_ids)

    traffic_list = query.values("stream_id").annotate(value=Sum("value"))
    traffic_dict = {}
    for traffic in traffic_list:
        traffic_dict[traffic["stream_id"]] = traffic["value"]

    return traffic_dict
```

The above functions fetches the count for multiple stream ids in a single fetch and then another function is called for each stream which does some simple arithmetic operations and rounding off to get the average traffic for each stream.

Reply:
OK, that seems fast enough to not be a big deal to do once when sending these events out.

Reply:
Just to confirm, do we want to then include this in the streams object as well, we currently include them in subscription objects only. Like in the `streams` field of register response and streams object sent as response to API request and also similar objects for default streams.

Reply:
@timabbott can you please confirm this? I think we should pass `stream_weekly_traffic` in the `streams` field of `register` response and in API endpoints like `get_streams_backend` and `get_stream_backend`. We do not use `streams` field of `register` response but other clients may do and would expect same fields to be present in the object which would be present for a stream received through stream creation event.

I am not sure about the stream obejcts sent in the response for default streams and default stream groups.

Reply:
"api design" is the best place to discuss this sort of thing -- I find it hard to keep up with the flood of PR comments on this sort of thing. I don't see much potential downside to moving it to the stream object other than performance, which I think we've confirmed is not a big deal above (but I think it having originally been not great before adding those analytics tables might be why it was this way in the first place? The Zephyr realm has the mirroring bot subscribed to ~20k streams and using the "get streams" endpoint, where a more expensive calculation might have been a problem -- though I think the fix there might be to just not send the traffic data for it). 

Please post in "api design" for visibility about this change but I think it's a fine path forward.